src/classes/ALLO_AllocationsDMLService.cls

```diff
@@ -0,0 +1,413 @@
+        Map<Id, List<Allocation__c>> backupPaymentAllocations;
+
+        Savepoint sp = Database.setSavepoint();
+
+        do {
+            // Clone existing allocation lists in case they need to be restored when rolling back
+            backupOppsAllocations = allocationsWrapper.oppsAllocations.clone();
+            backupPaymentAllocations = allocationsWrapper.paymentAllocations.clone();
+
+            dmlIteration++;
+            dmlErrors = false;
+
+            // Process Deletes
+            dmlErrors = (dmlErrors || processAllocationDeleteResults());
+
+            // Process Inserts
+            dmlErrors = (dmlErrors || processAllocationInsertResults());
+
+            // Process Updates
+            dmlErrors = (dmlErrors || processAllocationUpdateResults());
+
+            if (dmlErrors) {
+                // Rollback to a consistent state && clean the allocationResults for the next try of the loop
+                // Removing Allocations associated with Errored records via parentRecordId in errors Set
+                // Errors will remain to be logged
+                Database.rollback(sp);
+
+                // Revert Maps of existing Allocations to versions prior to DML Operation
+                allocationsWrapper.oppsAllocations = backupOppsAllocations;
+                allocationsWrapper.paymentAllocations = backupPaymentAllocations;
+
+                // Removes and Allocations associated with Parents with one or more DML Errors
+                removeAllDMLAssociatedWithErroredRecords();
+
+                // Insert DML Lists need special handling to clear id and read only timestamp fields from prior execution
+                removeIdAndTimestampsFromDMLList(allocationsWrapper.allocationsForInsert);
+            }
+        } while (dmlIteration <= numberOfDatabaseRetries && dmlErrors);
+
+        if (dmlErrors) { // Need to log errors for all Opportunities that don't already have errors since retries exhausted
+            addErrorsWhenRetriesExceeded();
+        }
+
+        // Convert any errors that were not explicitly DML errors here to RecordErrors
+        processFunctionalErrors();
+    }
+
+    /**
+     * @description Clears any private member variables of data that needs to be reset between processing
+     */
+    public void reset() {
+        recordIdsWithErrors = new Set<Id>();
+        allRecordIdsWithDML = new Set<Id>();
+        allocationsWrapper = null;
+        processExclusions = null;
+    }
+
+    /**
+     * @description Processes Allocation deletions, recording DML errors as Record Errors and adding
+     * Record Exclusions for ones that errored.  The ALLO_AllocationsWrapper Allocations Maps are updated removing
+     * Allocations from the Maps for successful DML.
+     * @return Boolean, true if there were any DML errors
+     */
+    @testVisible private Boolean processAllocationDeleteResults() {
+        List<Database.DeleteResult> deletes = null;
```

Comment:
How about defining this as `new List<Database.DeleteResult>()` and then you don't need line 233 to compare for not null?

Reply:
Let me see why I did that.  I don't remember if there was a good (or even a not good reason).

Reply:
I solved this by combining the two blocks, which eliminated the need for the null check.